lib/datadog/tracing/distributed/trace_context.rb

```diff
@@ -193,20 +193,22 @@ def serialize_origin(value)
 
         # Serialize `_dd.p.{key}` by first removing the `_dd.p.` prefix.
         # Then replacing invalid characters with `_`.
+        #
+        # Returns a new object if modified given `name` is a Hash key, thus always frozen.
         def serialize_tag_key(name)
           key = name.delete_prefix(Tracing::Metadata::Ext::Distributed::TAGS_PREFIX)
 
           # DEV: It's unlikely that characters will be out of range, as they mostly
           # DEV: come from Datadog-controlled sources.
           # DEV: Trying to `match?` is measurably faster than a `gsub` that does not match.
-          if INVALID_TAG_KEY_CHARS.match?(key)
-            key.gsub(INVALID_TAG_KEY_CHARS, '_')
-          else
-            key
-          end
+          key.gsub!(INVALID_TAG_KEY_CHARS, '_') if INVALID_TAG_KEY_CHARS.match?(key)
+
+          key
         end
```

Comment:
I'm confused by this change and the new comment.

Comment:

1. The comment states "Returns a new object if [...]". But I think we always return a new object -- because `String#delete_prefix` always returns a new object.
2. The comment mentions things being frozen, but the return of `String#delete_prefix` and `gsub!` is not frozen. Should we freeze the result?
3. I didn't understand the part about `name` being a Hash key -- can you clarify?

The change itself:

This seems to be an optimization -- mutate the string inline instead of creating a new one.

Have you considered going a bit further and replacing it with `key.gsub!(INVALID_TAG_KEY_CHARS, '_') || key`? Currently we seem to be "paying" for the regular expression matching twice -- once in `.match?` and again in `.gsub!`.

(Something similar seems to happen in `#serialize_origin` above)

Reply:
>     1. The comment states "Returns a new object if [...]". But I think we always return a new object -- because `String#delete_prefix` always returns a new object.

I think I copy/pasted from `serialize_tag_value`, which **conditionally** creates a new object. I changed the comment here given it's always a new object.

>     2. The comment mentions things being frozen, but the return of `String#delete_prefix` and `gsub!` is not frozen. Should we freeze the result?

We don't care if the result is frozen: it's only for our own consumption so leaving it either frozen or not is fine.

>     3. I didn't understand the part about `name` being a Hash key -- can you clarify?

String Hash keys are always frozen: 
```ruby
{'a' => 1}.keys[0].frozen?
=> true
```

And the parameter to this function (`name`) comes directly from a String Hash key.
Because we know it's frozen, we know we have to dup it to modify it.

> Have you considered going a bit further and replacing it with `key.gsub!(INVALID_TAG_KEY_CHARS, '_') || key`? Currently we seem to be "paying" for the regular expression matching twice -- once in `.match?` and again in `.gsub!`.

`key.gsub!(INVALID_TAG_KEY_CHARS, '_')` is unlikely to make changes to the String, given `INVALID_TAG_KEY_CHARS` are rare as the argument to this method is created by Datadog itself.
A `INVALID_TAG_KEY_CHARS.match?(key)` that returns false is much faster than a `key.gsub!(INVALID_TAG_KEY_CHARS, '_')` that doesn't change anything:

```ruby
require 'benchmark/ips'

REGEXP = /x/

Benchmark.ips do |x|
  x.report('gsub!') { 'a'.gsub!(REGEXP, 'b') }
  x.report('match?') { 'a'.match?(REGEXP) }
  x.compare!
end

# ruby 3.2.1
# Comparison:
#               match?:  7659457.3 i/s
#                gsub!:  3016072.8 i/s - 2.54x  slower
```

Reply:
Sounds good! Thanks for sharing the benchmarks -- as usual, it's much better to measure than to rely on intuition when talking about performance :)